File: api/src/importer.py
import json

class Team:
    def __init__(self):
        self.team_id = None
        self.rank = None
        self.kills = 0

def get_teams(match):
    teams = []

    for item in match["included"]:
        if item["type"] != "roster":
            continue

        team = Team()

        team.rank = item["attributes"]["stats"]["rank"]
        team.team_id = item["attributes"]["stats"]["teamId"]

        for participant in item["relationships"]["participants"]["data"]:
            for participant_item in match["included"]:
                if participant["id"] != participant_item["id"]:
                    continue

                team.kills = team.kills + participant_item["attributes"]["stats"]["kills"]
        
        teams.append(team)

    return teams

def get_opponents_conversion(toornament_match):
    conversion = {}

    for opponents in toornament_match["opponents"]:
        conversion[opponents["participant"]["custom_fields"]["team_id"]] = { 
            "number": opponents["number"],
            "name": opponents["participant"]["name"]
        }

    return conversion

def transform_teams_to_games(teams, toornament_match):
    game = {
        "status": "completed",
        "opponents": []
    }

    conversion = get_opponents_conversion(toornament_match)

    for team in teams:
        if str(team.team_id) in conversion:
            game["opponents"].append({
                "number": conversion[str(team.team_id)]["number"],
                "properties": {
                    "ingame_rank": team.rank,
                    "kills": team.kills
                }
            })

            print({
                "number": conversion[str(team.team_id)]["number"],
                "properties": {
                    "ingame_rank": team.rank,
                    "kills": team.kills
                }
            })

    return game

def preview_teams(teams, toornament_match):
    game = {
        "teams": []
    }

    conversion = get_opponents_conversion(toornament_match)

    for team in teams:
        if str(team.team_id) in conversion:
            game["teams"].append({
                "team": conversion[str(team.team_id)],
                "properties": {
                    "ingame_rank": team.rank,
                    "kills": team.kills
                }
            })

            print({
                "team": conversion[str(team.team_id)],
                "properties": {
                    "ingame_rank": team.rank,
                    "kills": team.kills
                }
            })

    return game

File: api/src/service/importer.rs
use super::{
    pubg,
    toornament
};

use std::collections::HashMap;

#[derive(Serialize, Debug)]
struct Team {
    team_id: i64,
    rank: i64,
    kills: i64
}

#[derive(Serialize, Debug, Clone)]
pub struct ConversionTeam {
    pub name: String,
    pub number: i64
}

fn get_opponents_conversion(toornament_match: &toornament::Match) -> HashMap<String, ConversionTeam> {
    let mut conversions = HashMap::new();

    for opponent in &toornament_match.opponents {
        if let Some(participant) = &opponent.participant {
            if let Some(team_id) = &participant.custom_fields.team_id {
                conversions.insert(
                    team_id.clone(),
                    ConversionTeam {
                        name: participant.name.clone(),
                        number: opponent.number.clone()
                    }
                );
            }
        }
    }

    conversions
}

fn get_teams(pubg_match: &pubg::MatchResponse) -> Vec<Team> {
    let mut teams = Vec::new();

    for include in &pubg_match.included {
        if let pubg::Include::Roster(roster) = include {
            let mut team = Team {
                team_id: roster.attributes.stats.team_id.clone(),
                rank: roster.attributes.stats.rank.clone(),
                kills: 0
            };

            for participant in &roster.relationships.participants.data {
                for item in &pubg_match.included {
                    if let pubg::Include::Participant(participant_item) = item {
                        if participant.id == participant_item.id {
                            team.kills += participant_item.attributes.stats.kills;
                        }
                    }
                }
            }

            teams.push(team);
        }
    }
    
    teams
}

#[derive(Serialize, Debug)]
pub struct PreviewProperty {
    pub ingame_rank: Option<i64>,
    pub kills: Option<i64>
}

#[derive(Serialize, Debug)]
pub struct PreviewTeam {
    pub team: ConversionTeam,
    pub properties: PreviewProperty 
}

#[derive(Serialize, Debug)]
pub struct Preview {
    pub teams: Vec<PreviewTeam>
}

pub fn get_preview(toornament_match: toornament::Match, pubg_match: pubg::MatchResponse) -> Preview {
    let teams = get_teams(&pubg_match);
    let conversions = get_opponents_conversion(&toornament_match);
    let mut preview_teams: Vec<PreviewTeam> = Vec::new();

    for team in teams {
        if let Some(conversion) = conversions.get(&team.team_id.to_string()) {
            preview_teams.push(PreviewTeam {
                team: conversion.clone(),
                properties: PreviewProperty {
                    ingame_rank: Some(team.rank),
                    kills: Some(team.kills)
                }
            }); 
        }
    }

    Preview {
        teams: preview_teams
    }
}

#[derive(Serialize)]
pub struct GameOpponent {
    pub number: i64,
    pub properties: PreviewProperty
}

#[derive(Serialize)]
pub struct Game {
    pub status: String,
    pub opponents: Vec<GameOpponent>
}

pub fn transform_teams_to_game(toornament_match: toornament::Match, pubg_match: pubg::MatchResponse) -> Game {
    let teams = get_teams(&pubg_match);
    let conversions = get_opponents_conversion(&toornament_match);
    let mut opponents: Vec<GameOpponent> = Vec::new();

    for team in teams {
        if let Some(conversion) = conversions.get(&team.team_id.to_string()) {
            opponents.push(GameOpponent {
                number: conversion.number.clone(),
                properties: PreviewProperty {
                    ingame_rank: Some(team.rank),
                    kills: Some(team.kills)
                }
            }); 
        }
    }

    for team in toornament_match.opponents {
        let mut finded = false;

        for opponent in &opponents {
            if opponent.number == team.number {
                finded = true;

                break;
            }
        }

        if finded == false {
            opponents.push(GameOpponent {
                number: team.number.clone(),
                properties: PreviewProperty {
                    ingame_rank: None,
                    kills: None
                }
            });
        }
    }

    Game {
        status: "completed".to_string(),
        opponents
    }
}

